File: src/main/java/december_2024/MaxProduct.java
package december_2024;

public class MaxProduct {

    public static void main(String[] args) {
        int[] nums = {1, 2, -3, 4};
//        int[] nums = {2, 3, -2, 4};
//        int[] nums = {-2, 0, -1};

        MaxProduct maxProduct = new MaxProduct();
        int result = maxProduct.maxProduct(nums);
        System.out.println(result);
    }

    public int maxProduct(int[] nums) {
        int res = Integer.MIN_VALUE;
        for (int n : nums) {
            res = Math.max(res, n);
        }
        int curMin = 1;
        int curMax = 1;

        for (int n : nums) {
            int temp = curMax * n;
            curMax = Math.max(n * curMax, Math.max(n * curMin, n));
            curMin = Math.min(temp, Math.min(n * curMin, n));
            res = Math.max(res, curMax);
        }
        return res;
    }

}

File: python/max_product.py
class Solution(object):
    # O(n) time | O(1) space
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = nums[0]
        curMin = 1
        curMax = 1

        for num in nums:
            temp = num * curMax
            curMax = max(num, num * curMax, num * curMin)
            curMin = min(num, num * curMin, temp)

            res = max(res, curMax)
        
        return res
    
solution = Solution()
print(solution.maxProduct([2,3,-2,4]))